src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvokerProvider.cs

```diff
@@ -239,6 +245,30 @@ internal List<Func<IRazorPage>> GetViewStartFactories(CompiledPageActionDescript
             return results;
         }
 
+        private PageHandlerBinderDelegate[] GetHandlerBinders(CompiledPageActionDescriptor actionDescriptor)
+        {
+            if (actionDescriptor.HandlerMethods == null ||
+                actionDescriptor.HandlerMethods.Count == 0 ||
+                actionDescriptor.HandlerMethods.All(handler => handler.Parameters.Count == 0))
+            {
+                return Array.Empty<PageHandlerBinderDelegate>();
+            }
```

Comment:
This seems like unneeded complexity. Also It looks like that third case will trip your assert at https://github.com/aspnet/Mvc/pull/7369/files#diff-d8384a52abde5770b0be9183b6b6c294R181

Reply:
We skip over parameters if the handler doesn't have any - https://github.com/aspnet/Mvc/pull/7369/files#diff-d8384a52abde5770b0be9183b6b6c294R162. My hope was to avoid allocating arrays with null values when executors don't have any parameters to bind (which going by our templates would be the most common case)